### PR TITLE
introduce ACK runtime and types collection

### PR DESCRIPTION
**Pull request by jaypipes**
This set of three commits adds common ACK types and a runtime
system. There are two primary packages:

* `pkg/types` contains the core interfaces and type definitions that
   ACK service controllers use
* `pkg/runtime` contains concrete objects that provide a common
  controller, resource management and reconciler implementation
  along with all the glue code that ties the aws-sdk-go into the
  AWS service-specific resource managers.

Included is an example bookstore service that represents how the
generated controller scaffolding might look like.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: services/example/pkg/resource/book.go [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+// on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+// express or implied. See the License for the specific language governing
+// permissions and limitations under the License.
+
+package resource
+
+// bookResource implements the `aws-service-operator-k8s/pkg/types.AWSResource`
+// interface
+type bookResource struct {
+}
```
Comment (jaypipes) [inline services/example/pkg/resource/book.go:19]: Note for reviewers: eventually this object will *wrap* the runtime.Object type that matches it (in the case here, that would be `services/example/apis/v1alpha1/book.go:Book`) and add some other metadata about relationships between the object and other objects.

File: pkg/runtime/README.md [hunk 2]
```diff
@@ -0,0 +1,54 @@
+# `pkg/runtime`
+
+This package contains a set of concrete structs and helper functions that
```
Comment (mhausenblas) [inline pkg/runtime/README.md:3]: "concrete structs" as opposed to?

File: pkg/runtime/README.md [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+provide a common controller *implementation* for an AWS service.
+
+The top-level container struct in the package is `ServiceController`. A single
+instance of `ServiceController` is created (using `NewServiceController()`)
+from the `cmd/controller/main.go` file that contains the ACK controller
+entrypoint for a specific AWS service API.
+
+`ServiceController` primarily serves as a way to glue the upstream
```
Comment (mhausenblas) [inline pkg/runtime/README.md:11]: upstream?

File: pkg/runtime/README.md [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+
+But how does the `ServiceController` know what kinds of CRs that it will
+handle?
+
+There is a `ServiceController.WithResourceManagerFactories()` method
+that sets the `ServiceController`'s collection of objects that implement the
+`types.AWSResourceManagerFactory` interface.
+
+These resource manager factories *produce* objects that implement the
+`types.AWSResourceManager` interface, which is basic CRUD+L operations for a
```
Comment (mhausenblas) [inline pkg/runtime/README.md:37]: not sure if CRUD+L is widely known outside of AWS

**Pull request merged**